### PR TITLE
Incorporated email or phone sign in/sign up functionality, added vali…

### PR DESCRIPTION
**Pull request by Usamasquare**
…dations on username, phone and password

### CODE REVIEW

File: app/assets/stylesheets/application.css [hunk 1]
```diff
@@ -13,3 +13,19 @@
  *= require_tree .
  *= require_self
  */
+@import “bootstrap”;
```
Comment (adam-wali) [inline app/assets/stylesheets/application.css]: is this working?

File: app/assets/stylesheets/application.css [hunk 2]
```diff
@@ -13,3 +13,19 @@
...
+.bg{
+  background-color:#a6c9e1;
+}
+.field_width{
+  width: 240px;
+}
+.section{
+  background-color: white;
+  margin: 0px 390px 1px 390px;
+  width: 60%;
```
Comment (adam-wali) [inline app/assets/stylesheets/application.css:34]: Let's update it to: and verify if it is working (no errors on chrome console)
```
@import 'bootstrap';

.center_div {
  margin: 0 auto;
  width: 20%;
}

.bg {
  background-color: #a6c9e1;
}

.field_width {
  width: 240px;
}

.section {
  background-color: white;
  margin: 0 390px 1px 390px;
  width: 60%;
}
```

File: app/controllers/users/confirmations_controller.rb [hunk 3]
```diff
@@ -0,0 +1,30 @@
+# frozen_string_literal: true
```
Comment (adam-wali) [inline app/controllers/users/confirmations_controller.rb]: remove this file if you aren't overriding any method

File: app/controllers/users/omniauth_callbacks_controller.rb [hunk 4]
```diff
@@ -0,0 +1,30 @@
+# frozen_string_literal: true
+
```
Comment (adam-wali) [inline app/controllers/users/omniauth_callbacks_controller.rb]: remove this file if you aren't overriding any method

File: app/controllers/users/passwords_controller.rb [hunk 5]
```diff
@@ -0,0 +1,34 @@
+# frozen_string_literal: true
+
```
Comment (adam-wali) [inline app/controllers/users/passwords_controller.rb]: remove this file if you aren't overriding any method

File: app/controllers/users/registrations_controller.rb [hunk 6]
```diff
@@ -0,0 +1,62 @@
+# frozen_string_literal: true
+
```
Comment (adam-wali) [inline app/controllers/users/registrations_controller.rb]: remove this file if you aren't overriding any method

File: app/controllers/users/unlocks_controller.rb [hunk 7]
```diff
@@ -0,0 +1,30 @@
+# frozen_string_literal: true
+
```
Comment (adam-wali) [inline app/controllers/users/unlocks_controller.rb]: remove this file if you aren't overriding any method

File: app/views/devise/registrations/edit.html.erb [hunk 8]
```diff
@@ -1,43 +1,54 @@
...
 
-  <div class="field">
+  <div class='field'>
     <%= f.label :password %> <i>(leave blank if you don't want to change it)</i><br />
-    <%= f.password_field :password, autocomplete: "new-password" %>
+    <%= f.password_field :password, autocomplete: 'new-password' %>
     <% if @minimum_password_length %>
       <br />
-      <em><%= @minimum_password_length %> characters minimum</em>
+      <em><%#= @minimum_password_length %> minimum 8 characters with at least one capital letter and a special character</em>
```
Comment (adam-wali) [inline app/views/devise/registrations/edit.html.erb]: remove commented code

File: app/views/devise/registrations/new.html.erb [hunk 9]
```diff
@@ -1,29 +1,44 @@
...
+            <%= f.password_field :password_confirmation, autocomplete: 'new-password' %>
+          </div>
+
+          <div class='actions'>
+            <%= f.submit 'Sign up' %>
+          </div>
+        <% end %>
 
-<%= render "devise/shared/links" %>
+    <%= render 'devise/shared/links' %>
```
Comment (adam-wali) [inline app/views/devise/registrations/new.html.erb]: indentation issue?
```suggestion
      <%= render 'devise/shared/links' %>
```

File: app/views/devise/registrations/new.html.erb [hunk 10]
```diff
@@ -1,29 +1,44 @@
...
+          </div>
 
-  <div class="field">
-    <%= f.label :password_confirmation %><br />
-    <%= f.password_field :password_confirmation, autocomplete: "new-password" %>
-  </div>
+          <div class='field'>
+            <b><%= f.label :password %></b>
+            <% if @minimum_password_length %>
+            <em>(<%#= @minimum_password_length %> minimum 8 characters with at least one capital letter and a special character)</em>
```
Comment (adam-wali) [inline app/views/devise/registrations/new.html.erb]: remove commented code

File: app/views/devise/sessions/new.html.erb [hunk 11]
```diff
@@ -2,8 +2,10 @@
 
 <%= form_for(resource, as: resource_name, url: session_path(resource_name)) do |f| %>
   <div class="field">
-    <%= f.label :email %><br />
-    <%= f.email_field :email, autofocus: true, autocomplete: "email" %>
+    <%#= f.label :email %><br />
+    <%#= f.email_field :email, autofocus: true, autocomplete: "email" %>
```
Comment (adam-wali) [inline app/views/devise/sessions/new.html.erb]: remove commented code

File: app/views/layouts/_header.html.erb [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+        <% end %>
+       
+        </ul>
+      <form class="d-flex">
+        <input class="form-control me-2" type="search" placeholder="Search" aria-label="Search">
+        <button class="btn btn-outline-success" type="submit">Search</button>
+      </form>
+    </div>
+  </div>
+</nav>
```
Comment (adam-wali) [inline app/views/layouts/_header.html.erb:36]: Indentation issues:
```
<nav class="navbar navbar-expand-lg navbar-dark bg-dark">
  <div class="container-fluid">
    <%= link_to 'Car Sell', root_path, class: "navbar-brand" %>
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>
    <div class="collapse navbar-collapse" id="navbarSupportedContent">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0">
        <li class="nav-item">
          <a class="nav-link active" aria-current="page" href="#">Home</a>
        </li>

        <% if user_signed_in? %>
          <li class="nav-item">
            <%= link_to 'Sign Out', destroy_user_session_path, method: :delete, class: 'nav-link' %>
          </li>
          <li class="nav-item">
            <%= link_to 'Edit Profile', edit_user_registration_path, class: 'nav-link' %>
          </li>
        <% else %>
          <li class="nav-item">
            <%= link_to 'Sign Up', new_user_registration_path, class: 'nav-link' %>
          </li>
          <li class="nav-item">
            <%= link_to 'Log In', new_user_session_path, class: 'nav-link' %>
          </li>
        <% end %>

      </ul>
      <form class="d-flex">
        <input class="form-control me-2" type="search" placeholder="Search" aria-label="Search">
        <button class="btn btn-outline-success" type="submit">Search</button>
      </form>
    </div>
  </div>
</nav>
```

File: app/views/layouts/_header.html.erb [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
+      <span class="navbar-toggler-icon"></span>
+    </button>
+    <div class="collapse navbar-collapse" id="navbarSupportedContent">
+      <ul class="navbar-nav me-auto mb-2 mb-lg-0">
+        <li class="nav-item">
+          <a class="nav-link active" aria-current="page" href="#">Home</a>
+        </li>
+        
+
```
Comment (adam-wali) [inline app/views/layouts/_header.html.erb]: remove extra  line

File: app/views/layouts/application.html.erb [hunk 14]
```diff
@@ -6,11 +6,15 @@
     <%= csrf_meta_tags %>
     <%= csp_meta_tag %>
 
+<link href="https://cdn.jsdelivr.net/npm/bootstrap@5.1.0/dist/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-KyZXEAg3QhqLMpG8r+8fhAXLRk2vvoC2f3B09zVXn8CA5QIVfZOJ3BCsw2P0p/We" crossorigin="anonymous">
+
+
```
Comment (adam-wali) [inline app/views/layouts/application.html.erb]: Remove this, after fixing scss file bootstrap should work, don't import it manually

File: app/assets/stylesheets/application.css (hunk 1 above)
Comment (Usamasquare) [reply app/assets/stylesheets/application.css]: no i m still working with bootstrap

File: app/models/user.rb [hunk 15]
```diff
@@ -2,5 +2,28 @@ class User < ApplicationRecord
...
+  # only allow letter, number, underscore and punctuation.
+  validates_format_of :phone, with: /^[0-9]{11}$/, :multiline => true
+  validates_format_of :password, with: /^(?=.*[A-Z])(?=.*[!@#\$%\^&\*])(?=.{8,})/, :multiline => true
+  validates_format_of :username, with: /^[-@.\/#&+\w\s]*$/, :multiline => true
+
+  attr_writer :login
+
+  def login
+    @login || self.phone || self.email
+  end
```
Comment (Usamasquare) [inline app/models/user.rb:17]: not sure. Used this code for devise overriding.

**Review by adam-wali — APPROVED** [review]

**Pull request merged**